File: flutter/lib/models/comment.dart
class Comment {
  final String id;
  final String name;
  final String email;
  final String movieId;
  final String text;
  final DateTime date;

  Comment({
    required this.id,
    required this.name,
    required this.email,
    required this.movieId,
    required this.text,
    required this.date,
  });

  factory Comment.fromJson(Map<String, dynamic> json) {
    return Comment(
      id: _extractId(json['_id']),
      name: json['name'] ?? '',
      email: json['email'] ?? '',
      movieId: _extractId(json['movie_id']),
      text: json['text'] ?? '',
      date: _parseDate(json['date']),
    );
  }

  static String _extractId(dynamic id) {
    if (id is String) return id;
    if (id is Map<String, dynamic> && id.containsKey('\$oid')) {
      return id['\$oid'] as String;
    }
    return id?.toString() ?? '';
  }

  static DateTime _parseDate(dynamic date) {
    if (date is String) {
      return DateTime.tryParse(date) ?? DateTime.now();
    }
    if (date is Map<String, dynamic> && date.containsKey('\$date')) {
      final dateInfo = date['\$date'];
      if (dateInfo is Map<String, dynamic> && dateInfo.containsKey('\$numberLong')) {
        final timestamp = int.tryParse(dateInfo['\$numberLong']) ?? 0;
        return DateTime.fromMillisecondsSinceEpoch(timestamp);
      }
    }
    if (date is int) {
      return DateTime.fromMillisecondsSinceEpoch(date);
    }
    return DateTime.now();
  }

  String get displayName => name.isNotEmpty ? name : 'Anonymous';
  
  String get formattedDate {
    final now = DateTime.now();
    final difference = now.difference(date);
    
    if (difference.inDays > 365) {
      return '${(difference.inDays / 365).floor()} year${difference.inDays > 365 * 2 ? 's' : ''} ago';
    } else if (difference.inDays > 30) {
      return '${(difference.inDays / 30).floor()} month${difference.inDays > 60 ? 's' : ''} ago';
    } else if (difference.inDays > 0) {
      return '${difference.inDays} day${difference.inDays > 1 ? 's' : ''} ago';
    } else if (difference.inHours > 0) {
      return '${difference.inHours} hour${difference.inHours > 1 ? 's' : ''} ago';
    } else if (difference.inMinutes > 0) {
      return '${difference.inMinutes} minute${difference.inMinutes > 1 ? 's' : ''} ago';
    } else {
      return 'Just now';
    }
  }
  
  String get displayText => text;
  
  bool get hasValidMovieId => movieId.isNotEmpty;
  
  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'name': name,
      'email': email,
      'movie_id': movieId,
      'text': text,
      'date': date.millisecondsSinceEpoch,
    };
  }
}

File: rn-expo/src/models/comment.ts
export class Comment {
  constructor(
    public id: string,
    public name: string,
    public email: string,
    public movieId: string,
    public text: string,
    public date: Date
  ) {}

  static fromJson(json: any): Comment {
    return new Comment(
      this.extractId(json._id),
      json.name || '',
      json.email || '',
      this.extractId(json.movie_id),
      json.text || '',
      this.parseDate(json.date)
    );
  }

  private static extractId(id: IdType): string {
    if (typeof id === 'string') return id;
    if (id && typeof id === 'object' && (id as { $oid?: string }).$oid) {
      return (id as { $oid: string }).$oid;
    }
    return id?.toString() || '';
  }

  private static parseDate(date: any): Date {
    if (typeof date === 'string') {
      return new Date(date);
    }
    if (date && typeof date === 'object' && date.$date) {
      const dateInfo = date.$date;
      if (dateInfo && typeof dateInfo === 'object' && dateInfo.$numberLong) {
        const timestamp = parseInt(dateInfo.$numberLong) || 0;
        return new Date(timestamp);
      }
    }
    if (typeof date === 'number') {
      return new Date(date);
    }
    return new Date();
  }

  get displayName(): string {
    return this.name || 'Anonymous';
  }

  get formattedDate(): string {
    const now = new Date();
    const difference = now.getTime() - this.date.getTime();
    const seconds = Math.floor(difference / 1000);
    const minutes = Math.floor(seconds / 60);
    const hours = Math.floor(minutes / 60);
    const days = Math.floor(hours / 24);
    const months = Math.floor(days / 30);
    const years = Math.floor(days / 365);

    if (years > 0) {
      return `${years} year${years > 1 ? 's' : ''} ago`;
    } else if (months > 0) {
      return `${months} month${months > 1 ? 's' : ''} ago`;
    } else if (days > 0) {
      return `${days} day${days > 1 ? 's' : ''} ago`;
    } else if (hours > 0) {
      return `${hours} hour${hours > 1 ? 's' : ''} ago`;
    } else if (minutes > 0) {
      return `${minutes} minute${minutes > 1 ? 's' : ''} ago`;
    } else {
      return 'Just now';
    }
  }

  get displayText(): string {
    return this.text;
  }

  get hasValidMovieId(): boolean {
    return this.movieId.length > 0;
  }

  toJson(): any {
    return {
      _id: this.id,
      name: this.name,
      email: this.email,
      movie_id: this.movieId,
      text: this.text,
      date: this.date.getTime(),
    };
  }
}